connectivity/FEATURE_BLE/source/common/ble_trace_helpers.cpp

```diff
@@ -516,4 +516,748 @@ const char* hci_opcode_to_string(uint16_t opcode)
 }
 #endif //MBED_CONF_MBED_TRACE_ENABLE
 
+const char *to_string(const UUID& uuid)
+{
+    static constexpr const  size_t  ROW_COUNT = 4;
+    static constexpr const char *HEX = "0123456789ABCDEF";
+    static constexpr const size_t HYPHENS_DELIMITER_COUNT = 4;
+
+    static char string_buffer[ROW_COUNT][UUID::LENGTH_OF_LONG_UUID + HYPHENS_DELIMITER_COUNT];
+    static uint8_t idx = 0;
+
+    ++idx;
+
+    if (idx == ROW_COUNT) {
+        idx= 0;
+    }
+
+    const uint8_t  length = uuid.getLen();
+    const char *reader  = (const char *) uuid.getBaseUUID() + length - 1;
+    char* writer  = string_buffer[idx];
+
+    if (length == UUID::LENGTH_OF_LONG_UUID) {
+        constexpr uint8_t format[] = {4, 2, 2, 2, 6};
+        for ( uint8_t char_count: format) {
+            for (size_t i = 0; i < char_count; ++i) {
+                *writer++ = HEX[(*reader >> 4) & 0xF];
+                *writer++ = HEX[(*reader--)    & 0xF];
+            }
+            *writer++ = '-';
+        }
+        *--writer = 0;
+    } else {
+        for (size_t i = 0; i < length; ++i) {
+            *writer++ = HEX[(*reader >> 4) & 0xF];
+            *writer++ = HEX[(*reader--)    & 0xF];
+        }
+        *writer = 0;
+    }
+
+    return string_buffer[idx];
+}
+
+const char *to_string(bool v)
```

Comment:
Quick question for my own personal C/C++ culture :

Compatibility issues with the existing code and bigger memory footprint put aside, would it be okay to use `std::string` instead of C-style string? Would a `std::string_view` with C++17 work as well? What are the pros and cons?

Something like:

```cpp
const std::string to_string(bool v)
{
    if (v) {
        return "true";
    } else {
        return "false";
    }
}
```

Reply:
`std::string` implies memory allocation of the string buffer. That is something we want to avoid; especially for constant data. `std::string_view` would be a better but in this case there is two issues: 
- It is part of C++17 while Mbed uses C++ 14. 
- Mbed trace does not understand `string_view` so we would have to use the `data()` member function everywhere. 

In the end, these helpers don't return non const or uninitialized memory and their output target `mbed_trace`. I think it is the right thing to use `const char*`.

Reply:
Thanks for the detailed answer! After posting, I did some experiment and using `std::string_view` vs `cont char*` uses the exact same memory footprint. Of course as you pointed out, you'll need the `data()` member function everywhere you need to pass a C-string, `printf()` will be okay but you'll get a compiler warning.

I'm wondering if in the code that we write and control, we should use the C-string style or move to `std::string_view` to keep up with the nice features. The `data()` function is not that bad and would only be used where we interface with mbed or legacy code.

Reply:
We take these decision depending on the benefits and expectations from the user base. The helpers in this PR are internal to the BLE code base and not exposed to our users. They are not meant to be manipulated, return constant value and just meant to be fed into a `tr_` trace. In that context, I'm not seeing much benefit for my team to recreate a C++14 version of `string_view` which main benefit is to be a compatible replacement for `std::string` and use it just for internal purposes. 

On the other hand, when it make sense for the users we don't hesitate to back port standard APIs. For example, `mbed::Span` was present in mbed OS even before standardization because it just make sense to avoid carrying around the `size` associated to an array.